### PR TITLE
Merkle circuit 

### PR DESCRIPTION
**Pull request by KiriosK**

### CODE REVIEW

File: primitive/zk_rollup/src/operator.rs [hunk 1]
```diff
@@ -1,45 +1,65 @@
+use std::vec;
```
Comment (ashWhiteHat) [inline primitive/zk_rollup/src/operator.rs:1]: can we use `sp_std::vec` instead?

File: primitive/zk_rollup/src/merkle_circuit.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    use crate::{
+        domain::{TransactionData, UserData},
+        operator::RollupOperator,
+        poseidon::Poseidon,
+    };
+
+    use super::BatchCircuit;
+
+    #[test]
+    fn batch_update() {
```
Comment (ashWhiteHat) [inline primitive/zk_rollup/src/merkle_circuit.rs:123]: rename to `zkrollup_test`

File: primitive/zk_rollup/src/merkle_circuit.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        let n = 15;
+        let label = b"verify";
+        let mut rng = StdRng::seed_from_u64(8349u64);
+        let mut pp = KzgParams::setup(n, BlsScalar::random(&mut rng));
+        const ACCOUNT_LIMIT: usize = 2;
+        const BATCH_SIZE: usize = 2;
+
+        // Create an operator and contract
```
Comment (ashWhiteHat) [inline primitive/zk_rollup/src/merkle_circuit.rs:131]: can you give a serial number for each sequence?

File: primitive/zk_rollup/src/merkle_circuit.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+        let alice = UserData::new(0, 10, alice_address);
+        let bob = UserData::new(1, 0, bob_address);
+
+        let deposit1 = TransactionData::new(alice_address, contract_address, 10)
+            .signed(alice_secret, &mut rng);
+        let deposit2 =
+            TransactionData::new(bob_address, contract_address, 0).signed(bob_secret, &mut rng);
+
+        let poseidon = Poseidon::<Fr, 2>::new();
```
Comment (ashWhiteHat) [inline primitive/zk_rollup/src/merkle_circuit.rs:151]: is this necessary?

File: primitive/zk_rollup/src/merkle_circuit.rs (hunk 3 above)
Comment (ashWhiteHat) [reply primitive/zk_rollup/src/merkle_circuit.rs:131]: zkrollup construction and users params construction should be upper part

File: primitive/zk_rollup/src/merkle_circuit/root.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    composer: &mut Builder<TatePairing>,
+    leaf: Fr,
+    root: Fr,
+    path: [(Fr, Fr); K],
+    path_pos: [u64; K],
+) -> Result<(), Error> {
+    let real_root = composer.append_witness(root);
+
+    let root = calculate_root(composer, leaf, path, path_pos)?;
+    composer.assert_equal(root, real_root);
```
Comment (ashWhiteHat) [inline primitive/zk_rollup/src/merkle_circuit/root.rs:115]: what's real_root?
new_root or precomputed_root would be better

File: primitive/zk_rollup/src/merkle_circuit/root.rs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    Ok(())
+}
+
+impl<const K: usize> Circuit<TatePairing> for MerkleMembershipCircuit<K> {
+    fn circuit(&self, composer: &mut Builder<TatePairing>) -> Result<(), Error> {
+        check_membership(composer, self.leaf, self.root, self.path, self.path_pos)
```
Comment (ashWhiteHat) [inline primitive/zk_rollup/src/merkle_circuit/root.rs:121]: You can define the check_membership logic here not to foward

File: primitive/zk_rollup/src/redjubjub_circuit.rs [hunk 7]
```diff
@@ -29,35 +29,44 @@ impl RedJubjubCircuit {
...
-        let s_bp_neg = composer.component_mul_point(neg, s_bp);
-        let s_bp_neg_r = composer.component_add_point(s_bp_neg, r);
-        let s_bp_neg_r_hash_pub_key = composer.component_add_point(s_bp_neg_r, hash_pub_key);
-        let finalized =
-            composer.component_mul_point(sapling_redjubjub_cofactor, s_bp_neg_r_hash_pub_key);
-
-        composer.assert_equal_public_point(finalized, JubjubExtended::ADDITIVE_IDENTITY);
-
-        Ok(())
+        check_signature(composer, self.public_key, self.signature, self.msg_hash)
```
Comment (ashWhiteHat) [inline primitive/zk_rollup/src/redjubjub_circuit.rs:69]: You can define the check_signature logic here not to foward

**Review by ashWhiteHat — CHANGES_REQUESTED** [review]

File: primitive/zk_rollup/src/operator.rs (hunk 1 above)
Comment (KiriosK) [reply primitive/zk_rollup/src/operator.rs:1]: Probably, I will check.

File: primitive/zk_rollup/src/merkle_circuit.rs (hunk 2 above)
Comment (KiriosK) [reply primitive/zk_rollup/src/merkle_circuit.rs:123]: This is not a zkrollup test, but it's better to rename the file I guess and the test to `batch_circuit_test`
The zkrollup one is is in the `lib.rs`

File: primitive/zk_rollup/src/merkle_circuit/root.rs (hunk 6 above)
Comment (KiriosK) [reply primitive/zk_rollup/src/merkle_circuit/root.rs:121]: This circuit is made just for testing inner logic. The function is used in the `BatchCircuit` instead.

File: primitive/zk_rollup/src/redjubjub_circuit.rs (hunk 7 above)
Comment (KiriosK) [reply primitive/zk_rollup/src/redjubjub_circuit.rs:69]: Same with the previous

File: primitive/zk_rollup/src/merkle_circuit.rs (hunk 2 above)
Comment (ashWhiteHat) [reply primitive/zk_rollup/src/merkle_circuit.rs:123]: Exactly

File: primitive/zk_rollup/src/operator.rs (hunk 1 above)
Comment (KiriosK) [reply primitive/zk_rollup/src/operator.rs:1]: Made this crate `no_std`

File: primitive/zk_rollup/src/merkle_circuit.rs (hunk 4 above)
Comment (KiriosK) [reply primitive/zk_rollup/src/merkle_circuit.rs:151]: Removed

File: primitive/zk_rollup/src/merkle_circuit/root.rs (hunk 5 above)
Comment (KiriosK) [reply primitive/zk_rollup/src/merkle_circuit/root.rs:115]: Fixed

**Pull request merged**